northstar-tests/northstar-tests-derive/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+        let runtime = northstar_tests::runtime::Runtime::new().expect("Failed to start runtime");
+
+        // The test code within the async context
+        let body = async {
+            let runtime = runtime.start().await?;
+            let r: anyhow::Result<()> = #body;
+            runtime.shutdown().await?;
+            r
+        };
+
+        // Run the test body inside of the Tokio runtime
+        #[allow(clippy::expect_used)]
+        tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .worker_threads(2)
+            .build()
+            .expect("Failed building the Runtime")
+            .block_on(body)
+
+    };
+
+    let brace_token = input.block.brace_token;
+    input.block = syn::parse2(quote_spanned! {last_stmt_end_span=>
+        {
+            use std::os::unix::io::{IntoRawFd, AsRawFd};
+            use std::io::{BufRead, Seek};
+            use nix::sys::wait::{waitpid, WaitStatus};
+            use nix::unistd::{ForkResult, dup2};
+            use std::process::exit;
+
+            // Create a memfd for capturing stdout/stderr of the child (test) process
+            let mfd = memfd::MemfdOptions::default().create("io").unwrap();
+
+            match unsafe { nix::unistd::fork().expect("Fork failed") } {
+                ForkResult::Parent { child, .. } => {
+                    let result = waitpid(child, None).unwrap();
+
+                    // Copy stdout/stderr from the test to the parents stdout.
+                    let mut output = mfd.into_file();
+                    output.seek(std::io::SeekFrom::Start(0)).unwrap();
+                    // This needs do be done linewise, because a std::io::copy
+                    // to std::io::stdout results in conditionless printed output.
+                    std::io::BufReader::new(output).lines().for_each(|line| println!("{}", line.unwrap()));
+
+                    // Return depending on the child exit code.
+                    match result {
+                        WaitStatus::Exited(_, code) if code == 0 => Ok(()),
+                        s => Err(anyhow::anyhow!("Test failed with status: {:?}", s)),
+                    }
+                }
+                ForkResult::Child => {
+                    // Replace stout and stderr with the memfd.
+                    dup2(mfd.as_raw_fd(), 1).unwrap();
+                    dup2(mfd.into_raw_fd(), 2).unwrap();
+
+                    // Run the test body and exit depending on the result.
+                    match { #test } {
+                        Ok(_) => exit(0),
+                        Err(e) => {
+                            eprintln!("{}", e);
+                            exit(1);
+                        }
+                    }
+                }
+            }
+        }
+    })
+    .expect("Parsing failure");
+    input.block.brace_token = brace_token;
+
+    let result = quote! {
+        #[::core::prelude::v1::test]
+        #input
+    };
+
+    result.into()
+}
```

Comment:
really fancy! but not sure how much value this brings :confused: it looks like it is doing pretty much the same as the old macro. Maybe it would be nice for the future if the tests become more complicated and more control is needed for setting up the runtime.

Reply:
The main motivation is that I need something similar in a different project and used the infrastructure here for testing. The benefit today is that you get better error messages from the compiler if something is wrong within your test (and a fixed return type) and the removal of the `rusty_fork` dependecy (which hasn't been updated for two years).